code/modules/wiremod/components/list/format.dm

```diff
@@ -0,0 +1,113 @@
+/**
+ * # Format List Component
+ *
+ * Formats lists by replacing %n in format string with nth parameter.
+ * Alternative to the Concatenate component.
+ */
+/obj/item/circuit_component/format
+	display_name = "Format List"
+	desc = "A component that formats lists, replacing %n in the format string with corresponding nth list item."
+	category = "List"
+
+	var/static/regex/list_param_regex = new(@"%([0-9]+)", "g")
+	// Used to provide what src should be in the regex replace proc. Necessary due to terrible byond API.
+	var/static/obj/item/circuit_component/format/regex_context
+
+	/// The regex used to find a parameter.
+	var/regex/param_regex
+
+	var/datum/port/input/format_port
+	var/datum/port/input/param_list_port
+
+	// Range for entity tostring to work, same mechanic as the To String component
+	var/max_range = 7
+
+	/// The result from the output
+	var/datum/port/output/output
+	circuit_flags = CIRCUIT_FLAG_INPUT_SIGNAL|CIRCUIT_FLAG_OUTPUT_SIGNAL
+
+/obj/item/circuit_component/format/Initialize(mapload)
+	. = ..()
+	param_regex = list_param_regex
+
+/obj/item/circuit_component/format/proc/make_params_port()
+	param_list_port = add_input_port("Params", PORT_TYPE_LIST(PORT_TYPE_ANY))
+
+/obj/item/circuit_component/format/populate_ports()
+	format_port = add_input_port("Format", PORT_TYPE_STRING)
+	make_params_port()
+
+	output = add_output_port("Output", PORT_TYPE_STRING)
+
+/**
+ * Get an item from the list.
+ * Return null to indicate invalid index.
+ * Arguments:
+ * * param_list - The resolved list.
+ * * index_string - The raw list index, as a string.
+ */
+/obj/item/circuit_component/format/proc/get_list_item(list/param_list, index_string)
+	var/index = text2num(index_string)
+	if(index < 1 || index > length(param_list))
+		return null
+
+	return param_list[index]
+
+/**
+ * Replace %n with the actual param, as a string.
+ * Arguments:
+ * * match - The full %1 regex match. Unused.
+ * * index_string - Just the "1" of the %1 format, actually used.
+ */
+/obj/item/circuit_component/format/proc/process_param(match, index_string)
+	// The static regex_context var is what you'd expect src to be, but src is actually the regex instance.
+	var/list/param_list = regex_context.param_list_port.value
+	if(!islist(param_list))
+		return @"[NO LIST]"
+
+	var/value = regex_context.get_list_item(param_list, index_string)
+	if(value == null)
+		return @"[BAD INDEX]"
+
+	// If this is a datum or atom, it's likely wrapped in a weakref.
+	if(isweakref(value))
+		var/datum/weakref/weak_value = value
+		value = weak_value.resolve()
+
+	// Working with entities is constrained by range, just as with To String.
+	if(isatom(value))
+		var/turf/location = regex_context.get_location()
+		var/turf/target_location = get_turf(value)
+		if(target_location.z != location.z || get_dist(location, target_location) > regex_context.max_range)
+			return @"[OUT OF RANGE]"
+
+	return "[value]"
```

Comment:
it seems like you could just replace this with a global proc instead, where you can then pass in the param list directly

Reply:
So this proc's argument list is hardwired by how regexes work, so wherever the functionality lives, there has to be a proc with the signature (match, index_capture) that either contains it or calls it. The full functionality is also very specific to this one component's behavior, so it would end up being a global proc that's otherwise identical.

Reply:
yeah that's fine, it's just that you shouldn't be using procs like these where src isn't the item itself because that's going to end up in a real clusterfuck

Reply:
Again, I have no control over what's passed to the proc properly, it's hardwired by byond. Using the static var is suboptimal, but is the solution tgstation already utlitizes for this use of regex.Replace(). Yeah I'll make it a global proc though so it's not confusing.

Reply:
In order to pass the object being affected coherently, I ended up subclassing `/regex` so that the regex instance holds it, and the proc in question lives in the regex, meaning `src` points to the expected value and is used. Still slightly cursed, but `/regex` is inherently cursed.

Reply:
> Again, I have no control over what's passed to the proc properly, it's hardwired by byond. Using the static var is suboptimal, but is the solution tgstation already utlitizes for this use of regex.Replace(). Yeah I'll make it a global proc though so it's not confusing.

makes sense